File: app/src/main/java/com/kickstarter/services/apirequests/ProjectNotificationBody.java
package com.kickstarter.services.apirequests;

import android.os.Parcelable;

import com.kickstarter.libs.qualifiers.AutoGson;

import auto.parcel.AutoParcel;

@AutoGson
@AutoParcel
public abstract class ProjectNotificationBody implements Parcelable {
  public abstract boolean email();
  public abstract boolean mobile();

  @AutoParcel.Builder
  public abstract static class Builder {
    public abstract Builder email(boolean __);
    public abstract Builder mobile(boolean __);
    public abstract ProjectNotificationBody build();
  }

  public static Builder builder() {
    return new AutoParcel_ProjectNotificationBody.Builder();
  }

  public abstract Builder toBuilder();
}

File: app/src/main/java/com/kickstarter/services/apirequests/ProjectNotificationBody.kt
package com.kickstarter.services.apirequests

import android.os.Parcelable
import kotlinx.parcelize.Parcelize

@Parcelize
class ProjectNotificationBody private constructor(
    private val email: Boolean,
    private val mobile: Boolean
) :
    Parcelable {
    fun email() = this.email
    fun mobile() = this.mobile

    @Parcelize
    data class Builder(
        private var email: Boolean = false,
        private var mobile: Boolean = false,
    ) : Parcelable {
        fun email(email: Boolean) = apply { this.email = email }
        fun mobile(mobile: Boolean) = apply { this.mobile = mobile }

        fun build() = ProjectNotificationBody(
            email = email,
            mobile = mobile
        )
    }

    fun toBuilder() = Builder(
        email = email,
        mobile = mobile
    )

    companion object {
        @JvmStatic
        fun builder() = Builder()
    }

    override fun equals(obj: Any?): Boolean {
        var equals = super.equals(obj)
        if (obj is ProjectNotificationBody) {
            equals = email() == obj.email() &&
                mobile() == obj.mobile()
        }
        return equals
    }
}
